database/sqlite/sqlite_health.c

```diff
@@ -0,0 +1,958 @@
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+
+    param++;
+    rc = sqlite3_bind_text(res, 28, cfg->host_labels, -1, SQLITE_STATIC);
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+
+    if (cfg->p_db_lookup_after) {
+        param++;
+        rc = sqlite3_bind_text(res, 29, cfg->p_db_lookup_dimensions, -1, SQLITE_STATIC);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+
+        param++;
+        rc = sqlite3_bind_text(res, 30, cfg->p_db_lookup_method, -1, SQLITE_STATIC);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+
+        param++;
+        rc = sqlite3_bind_int(res, 31, cfg->p_db_lookup_options);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+
+        param++;
+        rc = sqlite3_bind_int(res, 32, cfg->p_db_lookup_after);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+
+        param++;
+        rc = sqlite3_bind_int(res, 33, cfg->p_db_lookup_before);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+    } else {
+        param++;
+        rc = sqlite3_bind_null(res, 29);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+        param++;
+        rc = sqlite3_bind_null(res, 30);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+        param++;
+        rc = sqlite3_bind_null(res, 31);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+        param++;
+        rc = sqlite3_bind_null(res, 32);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+        param++;
+        rc = sqlite3_bind_null(res, 33);
+        if (unlikely(rc != SQLITE_OK))
+            goto bind_fail;
+    }
+
+    param++;
+    rc = sqlite3_bind_int(res, 34, cfg->p_update_every);
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+
+    rc = execute_insert(res);
+    if (unlikely(rc != SQLITE_DONE))
+        error_report("Failed to store alert config, rc = %d", rc);
+
+    rc = sqlite3_reset(res);
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Failed to reset statement in alert hash_id store function, rc = %d", rc);
+
+    return 0;
+
+    bind_fail:
+    error_report("Failed to bind parameter %d to store alert hash_id, rc = %d", param, rc);
+    rc = sqlite3_reset(res);
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Failed to reset statement in alert hash_id store function, rc = %d", rc);
+    return 1;
+}
+
+int alert_hash_and_store_config(
+    uuid_t hash_id,
+    struct alert_config *cfg)
+{
+    EVP_MD_CTX *evpctx;
+    unsigned char hash_value[EVP_MAX_MD_SIZE];
+    unsigned int hash_len;
+    evpctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(evpctx, EVP_sha256(), NULL);
+
+    if (cfg->alarm) {
+        EVP_DigestUpdate(evpctx, cfg->alarm, strlen(cfg->alarm));
+    } else if (cfg->template_key)
+        EVP_DigestUpdate(evpctx, cfg->template_key, strlen(cfg->template_key));
+
+    if (cfg->os)
```

Comment:
I wonder if we should add "NULL" in case it is NULL... Lets consider 2 cases:
1. `cfg->os == "tralalala"  && cfg->host == NULL`
2. `cfg->os == NULL && cfg->host == "tralalala"` 
I think current code will result in same hash for both cases (dangerous)?

Reply:
I think the `else` case should be something like `EVP_DigestUpdate(evpctx, `\0`, 1);`

Reply:
Good point, thanks Timo!

It will fail, for example in this scenario of two "blocks" of configs:

```
 template: fping_host_latency
 families: *
       on: fping.latency
    class: Latency
     type: Other
component: Network
   lookup: average -10s unaligned of average
    units: ms
    every: 10s
    green: 500
     warn: $this > $green OR $max > $red
     crit: $this > $red
    delay: down 30m multiplier 1.5 max 2h
     info: average latency to the network host over the last 10 seconds
       to: sysadmin

 template: fping_host_latency
 families: *
       on: fping.latency
    class: Latency
     type: Other
component: Network
   lookup: average -10s unaligned of average
    units: ms
    every: 10s
      red: 500
     warn: $this > $green OR $max > $red
     crit: $this > $red
    delay: down 30m multiplier 1.5 max 2h
     info: average latency to the network host over the last 10 seconds
       to: sysadmin
```

Taking advantage of the ordering of calling, we need to hash something if the value is null. I tried `EVP_DigestUpdate(evpctx, '\0', 1);` and `EVP_DigestUpdate(evpctx, '\0', 0);` but won't work, so I went with hashing the word `NULL` instead in https://github.com/netdata/netdata/pull/11399/commits/8b0257421c90d5237b3af68955d9df55d78b500e

We might be able to use something else instead of `NULL`, or a different approach ?

Reply:
see my comments below (used IDE to comment on new code so it started new thread)

Reply:
EVP_DigestUpdate(evpctx, '\0', 0); yeah that wont work because we need pointer to 0 byte instead of 0 byte itself. See my example below or maybe `EVP_DigestUpdate(evpctx, "", 1);` should work? (empty string is just 0 byte)

Reply:
Maybe we can use a control character? e.g. `EVP_DigestUpdate(evpctx, "\n", 1);`  ... ?

Reply:
I strongly prefer 0 byte for following reasons:
- the hashing function itself can hash any data (it doesn't care if it is string or whatever)
- by using `EVP_DigestUpdate(evpctx, cfg->template_key, strlen(cfg->template_key))` you are hashing the strings without trailing 0 byte. By definition then the valid hashed data never contain 0 byte *(while ASCII and UTF8 encodings are used, in case of UTF16 we have bigger problems everywhere else in agent)*.
- therefore 0 byte can be safely used to mean missing member in hashed data (there will ever be clash)

Reply:
Ok thanks, pushed https://github.com/netdata/netdata/pull/11399/commits/8b43102095009381bde28939f4fa9aaa16c13dee